bgpd: BGP neighbor password change doesn't take effect with a a particular config on reboot

sudhanshukumar22 opened the pull request:
bgpd: BGP neighbor password change doesn't take effect with a a particular config on reboot

Description: when vrf add is received, add Vrf-name to the interface database. This is needed while binding the VRF interface to the BGP socket.
In this case, the global bgp config containing vrf is received before zebra sends vrf add message to BGP. When we receive the global bgp vrf message
first, vrf interface is not present in the interface database of BGP. So, while creating the global bgp socket with vrf, interface bind to the vrf interface fails. This creates a BGP socket without vrf binding(i.e. socket in default VRF). The socket is used by another BGP instance (if present for default VRF) to send out BGP packets.


Problem Description/Summary :
changing the neighbor password resets the session immediately but it doesn't use the password. It continues to operate without password.

Managed to recreate the issue with below reduced config.

Setup:
Sonic1------Sonic2

Test Steps:
1. Configure eBGP session between Sonic-1 and Sonic2 in default vrf using Peer-group.
2. Create a dummy non-default vrf BGP instance on Sonic1 device.
3. Save and reload Sonic1 device.
4. Once the device is up, try to configure 'password <string>' on Sonic1 BGP neighbor alone.
5. The eBGP session should go down and must not come up until matching password is configured on Sonic2 device. But it comes up. This is because BGP packets are sent using other BGP socket (which was created using dummy non-default vrf BGP instance where vrf binding to the BGP socket had failed).

Expected Behavior :
The eBGP session should go down and must not come up until matching password is configured on Sonic2 device.

Signed-off-by: sudhanshukumar22 sudhanshu.kumar@broadcom.com

polychaeta reviewed (dismissed):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (dismissed):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one

<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/40666c5c228a1f34246a40af19758f13/raw/426e75fb66a4bfa7df6c1b1db9cdc5f0e195cc08/cr_7432_1604148615.diff | git apply
```

</p>

<p>

```diff
diff --git a/bgpd/bgp_main.c b/bgpd/bgp_main.c
index 03f479033..2eb518c83 100644
--- a/bgpd/bgp_main.c
+++ b/bgpd/bgp_main.c
@@ -265,12 +265,13 @@ static int bgp_vrf_new(struct vrf *vrf)
 	if (BGP_DEBUG(zebra, ZEBRA))
 		zlog_debug("VRF Created: %s(%u)", vrf->name, vrf->vrf_id);
 
-	zlog_info("VRF Creation message received: %s(%u)", vrf->name, vrf->vrf_id);
+	zlog_info("VRF Creation message received: %s(%u)", vrf->name,
+		  vrf->vrf_id);
 	struct interface *ifp;
 	ifp = if_get_by_name(vrf->name, vrf->vrf_id);
-	if (ifp)
-	{
-	       zlog_info("VRF interface Created: %s(%u) ifindex %d", ifp->name, ifp->vrf_id, ifp->ifindex);
+	if (ifp) {
+		zlog_info("VRF interface Created: %s(%u) ifindex %d", ifp->name,
+			  ifp->vrf_id, ifp->ifindex);
 	}
 
 	return 0;
@@ -280,9 +281,10 @@ static int bgp_vrf_delete(struct vrf *vrf)
 {
 	if (BGP_DEBUG(zebra, ZEBRA))
 		zlog_debug("VRF Deletion: %s(%u)", vrf->name, vrf->vrf_id);
-	zlog_info("VRF Deletion message received: %s(%u)", vrf->name, vrf->vrf_id);
-	/* No need to delete vrf here as it will be deleted when BGP client receives 
-	ZEBRA_VRF_DELETE message from zebra */
+	zlog_info("VRF Deletion message received: %s(%u)", vrf->name,
+		  vrf->vrf_id);
+	/* No need to delete vrf here as it will be deleted when BGP client
+	receives ZEBRA_VRF_DELETE message from zebra */
 
 	return 0;
 }
diff --git a/bgpd/bgpd.c b/bgpd/bgpd.c
index 044587fe9..34f37d091 100644
--- a/bgpd/bgpd.c
+++ b/bgpd/bgpd.c
@@ -5656,8 +5656,8 @@ int peer_password_set(struct peer *peer, const char *password)
 		 */
 		if (BGP_PEER_SU_UNSPEC(peer))
 			return BGP_SUCCESS;
-		ret =  (bgp_md5_set(peer) >= 0) ? BGP_SUCCESS
-						: BGP_ERR_TCPSIG_FAILED;
+		ret = (bgp_md5_set(peer) >= 0) ? BGP_SUCCESS
+					       : BGP_ERR_TCPSIG_FAILED;
 		/* Send notification or reset peer depending on state. */
 		if (BGP_IS_VALID_STATE_FOR_NOTIF(peer->status))
 			bgp_notify_send(peer, BGP_NOTIFY_CEASE,

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (changes requested):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (changes requested):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

donaldsharp commented on bgpd/bgp_main.c:
```diff
@@ -249,13 +249,26 @@ static int bgp_vrf_new(struct vrf *vrf)
 	if (BGP_DEBUG(zebra, ZEBRA))
 		zlog_debug("VRF Created: %s(%u)", vrf->name, vrf->vrf_id);
 
+	zlog_info("VRF Creation message received: %s(%u)", vrf->name,
```
Please remove your debug messages

donaldsharp commented on bgpd/bgp_main.c:
```diff
@@ -249,13 +249,26 @@ static int bgp_vrf_new(struct vrf *vrf)
 {
 	if (BGP_DEBUG(zebra, ZEBRA))
 		zlog_debug("VRF Deletion: %s(%u)", vrf->name, vrf->vrf_id);
+	zlog_info("VRF Deletion message received: %s(%u)", vrf->name,
```
again please remove

donaldsharp commented on bgpd/bgpd.c:
```diff
@@ -5512,16 +5513,17 @@ int peer_password_set(struct peer *peer, const char *password)
 			XFREE(MTYPE_PEER_PASSWORD, member->password);
 		member->password = XSTRDUP(MTYPE_PEER_PASSWORD, password);
 
+		/* Attempt to install password on socket. */
```
What does this change accomplish?

sudhanshukumar22 replied:
when setting the password on the bgp socket, first we should apply the password on the socket and then, reset the session or send update to peer.

sudhanshukumar22 replied:
in the existing code, first we were sending update on the bgp socket to the neighbor (or sending reset to neighbor). Then we were updating the same socket with MD5 parameters.

donaldsharp commented on bgpd/bgpd.c:
```diff
@@ -5557,16 +5559,17 @@ int peer_password_unset(struct peer *peer)
 
 	/* Check if handling a regular peer. */
 	if (!CHECK_FLAG(peer->sflags, PEER_STATUS_GROUP)) {
+		/* Attempt to uninstall password on socket. */
```
What doese this change accomplish?

sudhanshukumar22 replied:
when unsetting the password on the bgp socket, first we should apply the password on the socket and then, reset the session or send update to peer.

donaldsharp reviewed (changes requested):
Please fix your signed-off-by line and also please rebase to latest master and force push.  We do not want master merge arrows from branches

ton31337 commented on bgpd/bgp_main.c:
```diff
@@ -265,13 +265,26 @@ static int bgp_vrf_new(struct vrf *vrf)
+	zlog_info("VRF Creation message received: %s(%u)", vrf->name,
+		  vrf->vrf_id);
+	struct interface *ifp;
+	ifp = if_get_by_name(vrf->name, vrf->vrf_id);
```
Missing a blank line after declarations

polychaeta reviewed (changes requested):
Thanks for your contribution to FRR!

* One of your commits has a missing or badly formatted `Signed-off-by` line; we can't accept your contribution until all of your commits have one

If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

sudhanshukumar22 commented:
> Please fix your signed-off-by line and also please rebase to latest master and force push. We do not want master merge arrows from branches

I have changed the signed-off-by message and rebased to latest master. Let me know, if anything else is needed.

sudhanshukumar22 commented:
I accidently closed the pull request. I am reopening it. @donaldsharp : Please approve this commit as I have addressed all your review comments.

ton31337 commented on bgpd/bgp_main.c:
```diff
@@ -265,6 +265,14 @@ static int bgp_vrf_new(struct vrf *vrf)
+
+	ifp = if_get_by_name(vrf->name, vrf->vrf_id);
+	if (ifp) {
+		zlog_info("VRF interface Created: %s(%u) ifindex %d", ifp->name,
```
zlog_debug here as well

ton31337 commented:
Can this be tested?

sudhanshukumar22 commented:
> Can this be tested?

I have tested it in our box. The steps for this are mentioned in the description section.

sudhanshukumar22 commented:
> > Can this be tested?
> 
> I have tested it in our box. The steps for this are mentioned in the description section.

Do you want me to write a testcase for this and add it here ?

ton31337 commented:
>Do you want me to write a testcase for this and add it here ?

Yes. In `tests/topotests`.

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/d318f30633c591e2ce476241308e81a1/raw/f41b8565b8838111ade98ba7f751cf78cbfa5d55/cr_7432_1604398839.diff | git apply
```

</p>

<p>

```diff
diff --git a/bgpd/bgp_main.c b/bgpd/bgp_main.c
index 08a850314..522e0d094 100644
--- a/bgpd/bgp_main.c
+++ b/bgpd/bgp_main.c
@@ -269,8 +269,8 @@ static int bgp_vrf_new(struct vrf *vrf)
 
 	ifp = if_get_by_name(vrf->name, vrf->vrf_id);
 	if (ifp) {
-		zlog_debug("VRF interface Created: %s(%u) ifindex %d", ifp->name,
-			  ifp->vrf_id, ifp->ifindex);
+		zlog_debug("VRF interface Created: %s(%u) ifindex %d",
+			   ifp->name, ifp->vrf_id, ifp->ifindex);
 	}
 
 	return 0;

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/9da2394a7275615d602cb48c52b6936d/raw/d9112189ed50db3dca99324db2976ea17fc4ac5e/cr_7432_1610706158.diff | git apply
```

</p>

<p>

```diff
diff --git a/bgpd/bgp_main.c b/bgpd/bgp_main.c
index d4e53b7d0..37a9bf0db 100644
--- a/bgpd/bgp_main.c
+++ b/bgpd/bgp_main.c
@@ -272,8 +272,8 @@ static int bgp_vrf_new(struct vrf *vrf)
 
 	ifp = if_get_by_name(vrf->name, vrf->vrf_id);
 	if (ifp) {
-		zlog_debug("VRF interface Created: %s(%u) ifindex %d", ifp->name,
-			  ifp->vrf_id, ifp->ifindex);
+		zlog_debug("VRF interface Created: %s(%u) ifindex %d",
+			   ifp->name, ifp->vrf_id, ifp->ifindex);
 	}
 
 	return 0;

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/83fb3b7a750d2210589fa415772de053/raw/d9112189ed50db3dca99324db2976ea17fc4ac5e/cr_7432_1610706327.diff | git apply
```

</p>

<p>

```diff
diff --git a/bgpd/bgp_main.c b/bgpd/bgp_main.c
index d4e53b7d0..37a9bf0db 100644
--- a/bgpd/bgp_main.c
+++ b/bgpd/bgp_main.c
@@ -272,8 +272,8 @@ static int bgp_vrf_new(struct vrf *vrf)
 
 	ifp = if_get_by_name(vrf->name, vrf->vrf_id);
 	if (ifp) {
-		zlog_debug("VRF interface Created: %s(%u) ifindex %d", ifp->name,
-			  ifp->vrf_id, ifp->ifindex);
+		zlog_debug("VRF interface Created: %s(%u) ifindex %d",
+			   ifp->name, ifp->vrf_id, ifp->ifindex);
 	}
 
 	return 0;

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

ton31337 commented:
Can we have a topotest covering this behavior?

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/4c654aff6336a56470fbde0dcc3438b5/raw/cbf9c0ca1b41678278f5e7d4334a86f1a2e6458c/cr_7432_1611721788.diff | git apply
```

</p>

<p>

```diff
diff --git a/bgpd/bgp_main.c b/bgpd/bgp_main.c
index 141bd1ff1..a5871942a 100644
--- a/bgpd/bgp_main.c
+++ b/bgpd/bgp_main.c
@@ -273,8 +273,8 @@ static int bgp_vrf_new(struct vrf *vrf)
 
 	ifp = if_get_by_name(vrf->name, vrf->vrf_id);
 	if (ifp) {
-		zlog_debug("VRF interface Created: %s(%u) ifindex %d", ifp->name,
-			  ifp->vrf_id, ifp->ifindex);
+		zlog_debug("VRF interface Created: %s(%u) ifindex %d",
+			   ifp->name, ifp->vrf_id, ifp->ifindex);
 	}
 
 	return 0;

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

sudhanshukumar22 commented:
@ton31337 : is it possible to know why BGP has crashed here in the testcase ?

ton31337 commented:
Backtrace is:

```
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(vrf_get+0xc3) [0x7fee49972493]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(routing_control_plane_protocols_control_plane_protocol_create+0xcd) [0x7fee4998cecd]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(+0x878b5) [0x7fee499448b5]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(+0x87f8f) [0x7fee49944f8f]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(nb_candidate_commit_apply+0x33) [0x7fee49945283]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(nb_candidate_commit+0x9e) [0x7fee499453ce]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(+0x88839) [0x7fee49945839]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(nb_cli_apply_changes+0x5ec) [0x7fee4994867c]
BGP: /usr/lib/frr/bgpd(+0x1939e2) [0x55e145d939e2]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(+0x51f1d) [0x7fee4990ef1d]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(command_config_read_one_line+0x41) [0x7fee49911141]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(config_from_file+0x4d) [0x7fee4991137d]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(vty_read_config+0x159) [0x7fee49977b89]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(+0x76c15) [0x7fee49933c15]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(thread_call+0x60) [0x7fee4996ed80]
BGP: /usr/lib/x86_64-linux-gnu/frr/libfrr.so.0(frr_run+0xe8) [0x7fee499340b8]
BGP: /usr/lib/frr/bgpd(main+0x3ce) [0x55e145d0991e]
BGP: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0) [0x7fee48bda840]
BGP: /usr/lib/frr/bgpd(_start+0x29) [0x55e145d0bec9]
BGP: in thread frr_config_read_in scheduled from lib/libfrr.c:984 frr_config_fork()
```

I believe you could replicate the same in your environment for all those failing tests.

sudhanshukumar22 commented:
Thanks @ton31337 : Can you tell me where did you get this log from ?
I will run it this testcase  test_bfd_connection() in my setup.

ton31337 commented:
For example: https://ci1.netdef.org/browse/FRR-FRRPULLREQ-TP1U1604I386-17166

sudhanshukumar22 commented:
@mwinter-osr  , @ton31337 I am able to reproduce the issue using " py.test -s bfd-vrf-topo1/test_bfd_vrf_topo1.py ". It is signal 6, indicating a heap corruption.
Can I do gdb on bgp before starting the testcase ?

sudhanshukumar22 commented:
2021/02/22 09:40:12 BGP: bgpd 7.7-dev-my-manual-build starting: vty@2605, bgp@<all>:179
BGP: Received signal 6 at 1613986812 (si_addr 0x6f00000714, PC 0x7f37fa8f9438); aborting...
BGP: /usr/lib/libfrr.so.0(zlog_backtrace_sigsafe+0x57) [0x7f37fb64a887]
BGP: /usr/lib/libfrr.so.0(zlog_signal+0xdd) [0x7f37fb64aa7d]
BGP: /usr/lib/libfrr.so.0(+0xaaa35) [0x7f37fb671a35]
BGP: /lib/x86_64-linux-gnu/libpthread.so.0(+0x11390) [0x7f37fac9f390]
BGP: /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x38) [0x7f37fa8f9438]
BGP: /lib/x86_64-linux-gnu/libc.so.6(abort+0x16a) [0x7f37fa8fb03a]
BGP: /lib/x86_64-linux-gnu/libc.so.6(+0x2dbe7) [0x7f37fa8f1be7]
BGP: /lib/x86_64-linux-gnu/libc.so.6(+0x2dc92) [0x7f37fa8f1c92]
BGP: /usr/lib/libfrr.so.0(+0x77bc2) [0x7f37fb63ebc2]
BGP: /usr/lib/libfrr.so.0(if_create_name+0x1f) [0x7f37fb63ebef]
BGP: /usr/lib/frr/bgpd() [0x46f19b]
BGP: /usr/lib/libfrr.so.0(vrf_get+0xc3) [0x7f37fb6847e3]
BGP: /usr/lib/libfrr.so.0(routing_control_plane_protocols_control_plane_protocol_create+0xcd) [0x7f37fb69f1fd]
BGP: /usr/lib/libfrr.so.0(+0x8fd75) [0x7f37fb656d75]
BGP: /usr/lib/libfrr.so.0(+0x9044f) [0x7f37fb65744f]
BGP: /usr/lib/libfrr.so.0(nb_candidate_commit_apply+0x33) [0x7f37fb657743]
BGP: /usr/lib/libfrr.so.0(nb_candidate_commit+0x9e) [0x7f37fb65788e]
BGP: /usr/lib/libfrr.so.0(+0x90cf9) [0x7f37fb657cf9]
BGP: /usr/lib/libfrr.so.0(nb_cli_apply_changes+0x5ec) [0x7f37fb65ab3c]
BGP: /usr/lib/frr/bgpd() [0x4f1e09]
BGP: /usr/lib/libfrr.so.0(+0x5a3dd) [0x7f37fb6213dd]
BGP: /usr/lib/libfrr.so.0(command_config_read_one_line+0x41) [0x7f37fb623601]
BGP: /usr/lib/libfrr.so.0(config_from_file+0x4d) [0x7f37fb62383d]
BGP: /usr/lib/libfrr.so.0(vty_read_config+0x159) [0x7f37fb689ed9]
BGP: /usr/lib/libfrr.so.0(+0x7f0d5) [0x7f37fb6460d5]
BGP: /usr/lib/libfrr.so.0(thread_call+0x60) [0x7f37fb6810d0]
BGP: /usr/lib/libfrr.so.0(frr_run+0xe8) [0x7f37fb646578]
BGP: /usr/lib/frr/bgpd(main+0x38f) [0x46cb8f]
BGP: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0) [0x7f37fa8e4840]
BGP: /usr/lib/frr/bgpd(_start+0x29) [0x46efa9]
BGP: in thread frr_config_read_in scheduled from lib/libfrr.c:984 frr_config_fork()

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/9a1640e58678f270a353095e65fd9d8d/raw/848d0f29bb14f4c2ac8b61533a7b499585f687d0/cr_7432_1617614928.diff | git apply
```

</p>

<p>

```diff
diff --git a/bgpd/bgp_main.c b/bgpd/bgp_main.c
index 61952fab4..96c2e130d 100644
--- a/bgpd/bgp_main.c
+++ b/bgpd/bgp_main.c
@@ -279,7 +279,7 @@ static int bgp_vrf_new(struct vrf *vrf)
 	if (ifp) {
 		if (BGP_DEBUG(zebra, ZEBRA))
 			zlog_debug("VRF interface Created: %s(%u) ifindex %d",
-				ifp->name, ifp->vrf_id, ifp->ifindex);
+				   ifp->name, ifp->vrf_id, ifp->ifindex);
 	}
 
 	return 0;

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

sudhanshukumar22 commented:
@ton31337 : Please review this code.

sudhanshukumar22 commented:
@qlyoung @donaldsharp @ton31337 : Please review the changes. These are minor changes. I had discussed with @mwinter-osr , currently we don't have testcases for reboot scenarios in the mininet framework.

ton31337 commented on lib/vrf.c:
```diff
@@ -1006,7 +1006,7 @@ int vrf_bind(vrf_id_t vrf_id, int fd, const char *name)
 	 */
 	ifp = if_lookup_by_name(name, vrf_id);
-	if (!ifp)
+	if (!ifp || ifp->ifindex == 0)
```
I suggest using `IFINDEX_INTERNAL` for 0.

sudhanshukumar22 replied:
Done

ton31337 commented on bgpd/bgp_main.c:
```diff
@@ -273,6 +273,15 @@ static int bgp_vrf_new(struct vrf *vrf)
 	if (BGP_DEBUG(zebra, ZEBRA))
 		zlog_debug("VRF Created: %s(%u)", vrf->name, vrf->vrf_id);
 
+	struct interface *ifp;
```
Can you move to the top of the function?

sudhanshukumar22 replied:
Done

sudhanshukumar22 commented on bgpd/bgp_main.c:
```diff
@@ -273,6 +273,15 @@ static int bgp_vrf_new(struct vrf *vrf)
+	if (ifp) {
+		if (BGP_DEBUG(zebra, ZEBRA))
+			zlog_debug("VRF interface Created: %s(%u) ifindex %d",
+				   ifp->name, ifp->vrf_id, ifp->ifindex);
```
vrf->name and ifp->name is same here. We are already displaying ifp->name here.

sudhanshukumar22 commented:
@ton31337 : Please review and approve the changes.

idryzhov reviewed (changes requested):
This is not a correct approach to solving this problem. We should not create interfaces manually from BGP. The correct thing is to fix `vrf_bind` to rely on VRF existence instead of interface existence. https://github.com/FRRouting/frr/pull/8367 is going to fix this.

qlyoung reviewed (commented):
I'm pretty confused about how this fixes the problem.

The root problem is that the VRF doesn't exist and so config for a BGP instanced tied to a VRF ends up creating a socket in the default VRF. Any later incorrect behavior that manifests because of this needs to be treated at the source.

Based on how I'm reading this patch, it looks like all we do here is make a interface that has the same name as the VRF we expect zebra to tell us about in short order. I fail to see how this treats the root of the problem.

Can you elaborate on how exactly this helps things? How does creating an internal interface structure make it so that the password (which is set via a sockopt call) ends up properly applied?

idryzhov commented:
@qlyoung the author already stated the vrf_bind PR fixes his use-case and he's going to dismiss this PR: https://github.com/FRRouting/frr/pull/8644/files#r630961190

qlyoung commented:
Excellent, @sudhanshukumar22 please close this

sudhanshukumar22 commented:
@qlyoung  @idryzhov , thanks, I have tested that my defect works with the fix provided in
https://github.com/FRRouting/frr/pull/8644/files#r630961190

I was waiting for his PR to be merged. Anyway, will close it now.
